xc/domain.ml

```diff
@@ -632,7 +633,61 @@ let create_channels ~xc uuid domid =
   debug "VM = %s; domid = %d; store evtchn = %d; console evtchn = %d" (Uuid.to_string uuid) domid store console;
   store, console
 
-let build_pre ~xc ~xs ~vcpus ~memory domid =
+let numa_hierarchy =
+  let open Xenctrlext in
+  let open Topology in
+  Lazy.from_fun (fun () ->
+      let distances = Xenctrlext.(with_xc numainfo).distances in
+      let cpu_to_node = Xenctrlext.(with_xc cputopoinfo) |> Array.map (fun t -> t.node) in
+      NUMA.v ~distances ~cpu_to_node)
+
+let numa_mutex = Mutex.create ()
+let numa_resources = ref None
+
+let numa_init () =
+  if !Xenopsd.numa_placement then begin
+    let host = Lazy.force numa_hierarchy in
+    let mem = Xenctrlext.(with_xc numainfo).memory in
+    D.debug "Host NUMA information: %s" (Fmt.to_to_string Topology.NUMA.pp_dump host);
+    Array.iteri (fun i m ->
+        let open Xenctrlext in
+        D.debug "NUMA node %d: %Ld/%Ld memory free" i m.memfree m.memsize
+      ) mem
+  end
+
+let numa_placement domid ~vcpus ~memory =
+  let open Xenctrlext in
+  let open Topology in
+  let hint = Mutex.execute numa_mutex (fun () ->
+      let host = Lazy.force numa_hierarchy in
+      let numa_meminfo = Xenctrlext.(with_xc numainfo).memory |> Array.to_list in
+      let nodes =
+        ListLabels.map2 (NUMA.nodes host |> List.of_seq) numa_meminfo ~f:(fun node m ->
+            NUMA.resource host node ~memory:m.memfree)
+      in
+      let vm = NUMARequest.v ~memory ~vcpus in
+      let nodea = match !numa_resources with
+        | None ->
+          let a = Array.of_list nodes in
+          numa_resources := Some a;
```

Comment:
This assignment could be done just before planning and avoid the repetition in the two branches.